### DISCUSSION TITLE
Simple binary build

### DISCUSSION THREAD

**Discussion by 6pac** [cesanta/mongoose#1408] [Ask questions, get answers] [answered]
I came here looking for a simple, single file webserver to serve static HTTP content. I realise that Mongoose can do way more than this, but it does also appear to be the best choice for this application, in which simplicity is the key element.

However it appears that it is no longer possible to download binaries, as covered here: https://stackoverflow.com/questions/29561734/i-cant-download-the-mongoose-exe/35801034#35801034
All the links to cesanta.com, say https://www.cesanta.com/products/binary, now give a 404.

I have checked out the HTTP tutorial at https://mongoose.ws/tutorials/http-server/, and it is certainly straightforward, however I'm not familiar with the C++ toolset and have no idea how to build the source.  
Could I request a short tutorial on building source for a very simple example (why not use the static HTTP server) on each of the platforms (Win, OSX, Linux).

At the moment I have been force to use an old version at https://www.softpedia.com/get/Internet/Servers/WEB-Servers/Mongoose.shtml, which works like a charm but I think you would agree is extremely non ideal from a security and availability perspective. However it's my only choice other than spending hours, perhaps days, attempting to get a build to work.

**Comment by cpq** [comment]
Sure!
- **Build for Linux or Mac** . This builds a binary called `example` which you can rename to `mongoose`
  ```sh
  $ git clone https://github.com/cesanta/mongoose
  $ cd mongoose/examples/http-server
  $ make example
  ```
- **Build for Windows on Windows** . Make sure to have MinGW installed. This builds a binary called `mongose.exe`
  ```sh
  $ git clone https://github.com/cesanta/mongoose
  $ cd mongoose/examples/http-server
  $ make mingw
  ```
- **Build for Windows on Mac/Linux** . Make sure to have Docker installed. This builds a binary called `mongose.exe`
  ```sh
  $ git clone https://github.com/cesanta/mongoose
  $ cd mongoose/examples/http-server
  $ make mongoose.exe
  ```

**Reply by zsinba** [reply]
make: *** No rule to make target `mongoose.exe'.  Stop.

**Reply by scaprile** [reply]
> make: *** No rule to make target `mongoose.exe'. Stop.

@zsinba 
Please don't necropost
Please don't thread-hijack
If you have an issue, open an issue, describe your scenario; describe what you do, what you expect, and what you get
With base on what you posted I have to rely on my divination powers: you seem to be a Windows user hoping to magically build something out of nothing and have run `make` on the root directory, or perhaps followed that old advice and don't understand what make is telling you: your rule does not exist.

> This thread is almost 2 years old, things do change
>**Please follow the guidelines in the [documentation](https://mongoose.ws/documentation)**

We have tons of [examples](https://github.com/cesanta/mongoose/tree/master/examples) and [tutorials](https://mongoose.ws/tutorials/) available, take advantage of them.

If you just want to build a web server, then build the http-server example: https://mongoose.ws/documentation/tutorials/http-server/
Shortcut:
```
git clone https://github.com/cesanta/mongoose
cd mongoose/examples/http-server
make
```
Valid for Linux/MacOS/Windows (mingw)

**Comment by 6pac** [comment]
Thanks Sergey, that was much simpler than I thought it would be. The tools are all built in.
I'm on a Mac, so this works straight off, but I will try to work out a procedure for Windows-only users, who are probably a large proportion of people wanting to do this. It may be possible using MinGW or equivalent.

**Comment by cpq** [comment]
I've described the procedure for windows. You can build `mongoose.exe` on your Mac. Just make sure you have Docker running.

**Comment by 6pac** [comment]
Sure, what I mean is for other people who don't have access to a Mac or Linux box and want to do the *build* on Windows. You have solved the issue for me, but I'm trying to solve the general problem here, especially since this is such a great all-round tool ;-)

**Comment by cpq** [comment]
I see, thank you.
You're correct, either MinGW or MSVC would work just fine, e.g. for MinGW:

```sh
i686-w64-mingw32-gcc ../../mongoose.c main.c -I../.. -W -Wall -DMG_ENABLE_IPV6=1 -DMG_ENABLE_LINES=1  -lws2_32 -o mongoose.exe
```

Let me know if that works. If yes, I'll add an extra `mingw` target to the example's Makefile, and add Windows-based build instruction to the answer.

**Reply by bing2008** [reply]
I use this comand build for windows succeeded.
I am using MinGW 6.3.0 and win10 , I just added "-D_POSIX_C_SOURCE=200000L" section to solve the "undefined reference to `fseeko'" error when i just use command above . [ref](https://stackoverflow.com/questions/40770673/implicit-declaration-of-function-fseeko) .  It's works fine now !  

`gcc ../../mongoose.c main.c -I../.. -W -Wall -DMG_ENABLE_IPV6=1 -DMG_ENABLE_LINES=1 -D_POSIX_C_SOURCE=200000L  -lws2_32 -o mongoose.exe`

**Comment by cpq** [comment]
Thank you @bing2008 . The `mingw` target with exactly that command was added to the Makefile, https://github.com/cesanta/mongoose/commit/d311e967fa4383ebcabb146fa826de87b91bb3a7

The answer above was modified to add a section `Build for Windows on Windows`.

Thanks again.

**Reply by Roph** [reply]
Is there some reason why you no longer provide a pre-built binary? I can do it myself, but like OP says, it forces many people who know nothing about command lines etc to look for potentially unsafe old binaries from random sites, or end up with straight malware downloads.

It seems very odd that you wouldn't provide a guaranteed safe binary straight from the source.

**Comment by 6pac** [comment]
I finally got back to this, but was happy to see it had been solved already.

Unfortunately for me, I had been assuming that my MinGW install that came as part of the Windows GIT client would do the compile. However it appears to be missing most of this tooling.

I just spent two hours trying to get something to work, unsuccessfully. I tried adding the tools to the MinGW I have, but I think it's just too cut down to do the job.
I went to the MinGW64 site (the MinGW.org site no longer exists, and nearly all of the info on the net is old and not longer relevant), and there are about 10 preinstalled packages to choose from. Again, I have no idea which one to choose and if it will coexist with my GIT tools (I have a large open source GIT repo set up locally and don't want to wreck the configuration).
I tried MSYS2 and it also doesn't seem to have make or gcc included. Plus it runs an MSYS2 terminal after install, but after that point I can't work out how to run it again: there's no start menu or context shortcut, and in the install folders there are about four different shells, in none of which make or gcc work, and I can't find any instructions as to even how I start the preferred MSYS2 shell.

So, @bing2008, could you provide any guidance as to which of the toolings you used at https://www.mingw-w64.org/downloads/

We need to provide a comprehensive, no assumptions process all the way from a vanilla Windows install to the final Mogoose build, including download and install of Mingw tooling.

**Reply by bing2008** [reply]
Sorry for the late, MinGW's install step confuse me too...

If use MinGW64 (not MinGW32, my last post was [MinGW32](https://sourceforge.net/projects/mingw/)), it can already worked with cpq's command line above: https://github.com/cesanta/mongoose/discussions/1408#discussioncomment-1692740 .

My steps using MinGW64 like this:
1. Use the "mingw-builds" link in https://www.mingw-w64.org/downloads/#mingw-builds . It links to Sourceforge download page and download the "mingw-w64-install.exe" installer. It can install MinGW64 automatically.
2. In win command line, use following lines (also can use in bat file ):
```
REM setup your mingw64 install folder to the PATH envment  
set "PATH=C:\Program Files (x86)\mingw-w64\i686-8.1.0-posix-dwarf-rt_v6-rev0\mingw32\bin;"

REM enter your mongoose source folder and http-server example folder
cd c:\mongoose-master\examples\http-server

REM build
i686-w64-mingw32-gcc.exe ../../mongoose.c main.c -I../.. -W -Wall -DMG_ENABLE_IPV6=1 -DMG_ENABLE_LINES=1 -lws2_32 -o mongoose.exe
```

3.If there's no error , it's done.

**Comment by 6pac** [comment]
So here's a complete start-to-end buiold procedure for Windows. Thanks due [here](https://superuser.com/questions/1294343/install-gcc-in-git-for-windows-bash-environment).  
It looks like the GIT Bash shell isn't up to the task and you have to do a full install of MINGW64. These instructions add the full MINGW64 package to a new folder and add it to the path, so it has the least interference potential to the existing GIT Bash  shell. 

- download and install 'Git for Windows'. Once installed, you will be able to open a bash shell at any folder by right clicking on the folder in Windows Explorer and selecting 'Git Bash here'.
- go to the MINGW home page [www.mingw-w64.org/downloads](https://www.mingw-w64.org/downloads/#mingw-builds) and follow the 'Mingw-builds' Github link
- there are a lot of slightly cryptic choices. The 'i686' versions are 32 bit, 'x86_64' are 32/64 bit. The 'posix' versions have more sophisticated threading capabilities, 'win32' is basic. 'seh' and 'sjlj' are different exception handling models. After a bit of research, the clear choice for a modern environment (circa 2022) is ```x86_64-12.2.0-release-posix-seh-rt_v10-rev0.7z```.
- download and unzip the file. You need to copy it to somewhere permanent on your system. I simply copied it to 'C:\Program Files\' (the folder unzupped as 'mingw64' so the full path was 'C:\Program Files\mingw64') but others created a folder for the version and build (you'd need to do that if you are going to run multiple versions or builds) and put it in that folder.
- you now need to add to the path the folder you just set up, using something like ```export PATH="/c/Program Files/mingw64/bin:$PATH"``` (note the forward slashes and the '/c/' at the start, and note that posts above have used the build-specific path, eg. ```set "PATH="/c/Program Files (x86)/mingw-w64/i686-8.1.0-posix-dwarf-rt_v6-rev0/mingw32/bin;```. Use whatever path you have set up.
- running the above command will add the folder to the path for that session only. To make in permanent, create a file called '.bashrc' in your user home folder (eg. in 'C:\Users\MyUserName\') and in a text editor paste the above command into the file. Next time you open the bash shell it will throw an error that a profile file is not present, but then it will create the profile file and from then on everything will work (this is the lazy way to do it!)
- just a note that the above all assumes you are running a bash shell - if you are trying to do this from a regular wndows command prompt, then you'd have to use regular windows notation (eg. 'C:\Program Files\mingw64\' for the path exports) and add the path using the Control Panel environment tools.

Finally, you are ready to download the repo and run the make. Open a GIT Bash shell where you want to store the repo, and:

    git clone https://github.com/cesanta/mongoose
    cd mongoose/examples/http-server
    mingw32-make mingw

or if that fails, use gcc directly:

    gcc ../../mongoose.c main.c -I../.. -W -Wall -DMG_ENABLE_IPV6=1 -DMG_ENABLE_LINES=1 -D_POSIX_C_SOURCE=200000L -lws2_32 -o mongoose.exe 

The ```which``` command in bash or the ```where``` command in Windows CMD are invaluable tools for tracking down any problems. ```which gcc``` in bash searches the PATH for the gcc and reports where it would be executed from. ```where gcc``` in CMD does the same but lists all occurrences. If either make or gcc are not working, these are your best tools to work out what in your path export has gone wrong. Note that the 'make' command does not appear to work out of the box in MINGW64. It appears the only version available is ```mingw32-make```. I assume this is for disambiguation reasons to do with 32/64 bit support.

**Reply by 6pac** [reply]
@ryuukk thanks for the mention of ```vcpkg``` but I thought the requirement of Visual Studio was too heavy for many users.

**Comment by cpq** [comment, marked as answer]
@6pac we've created an auto-build for the examples/http-server, with pre-built binaries for Windows, Mac and Linux: https://mongoose.ws/binary/

**Reply by 6pac** [reply]
+1

**Reply by hauk92** [reply]
@cpq it seems like the binaries were only build once. They were not updated after version 7.9.

**Reply by cpq** [reply]
@hauk92 thanks. Rebuilt with 7.12
